### PR TITLE
runtime/fcntl

### PR DESCRIPTION
**Pull request by leongross**
Rework of #4481

### CODE REVIEW

File: src/runtime/os_linux.go [hunk 1]
```diff
@@ -139,3 +139,12 @@ func hardwareRand() (n uint64, ok bool) {
 //
 //export getrandom
 func libc_getrandom(buf unsafe.Pointer, buflen uintptr, flags uint32) uint32
+
+// int fcntl(int fd, int cmd, int arg);
+//
+// export fcntl
+func libc_fcntl(fd int, cmd int, arg int) (ret int)
+
+func fcntl(fd int, cmd int, arg int) int {
```
Comment (aykevl) [inline src/runtime/os_linux.go]: This must match the signature in the syscall/unix package:

```go
func fcntl(fd int32, cmd int32, arg int32) (int32, int32)
```

File: src/runtime/os_linux.go [hunk 2]
```diff
@@ -139,3 +146,16 @@ func hardwareRand() (n uint64, ok bool) {
...
+
+// int fcntl(int fd, int cmd, int arg);
+//
+// export fcntl
+func libc_fcntl(fd int, cmd int, arg int) (ret int)
+
+func fcntl(fd int32, cmd int32, arg int32) (ret int32, errno int32) {
+	ret = int32(libc_fcntl(int(fd), int(cmd), int(arg)))
+	if ret < 0 {
+		return 0, int32(uintptr(*libc_errno_location()))
```
Comment (aykevl) [inline src/runtime/os_linux.go]: Why this double cast?

```suggestion
		return 0, *libc_errno_location()
```

File: src/runtime/os_linux.go [hunk 3]
```diff
@@ -139,3 +146,16 @@ func hardwareRand() (n uint64, ok bool) {
...
+//
+//export fcntl
+func libc_fcntl(fd int, cmd int, arg int) (ret int)
+
+func fcntl(fd int32, cmd int32, arg int32) (ret int32, errno int32) {
+	ret = int32(libc_fcntl(int(fd), int(cmd), int(arg)))
+	if ret < 0 {
+		return 0, int32(*libc_errno_location())
+	}
+	return ret, 0
```
Comment (aykevl) [inline src/runtime/os_linux.go]: This is not quite correct. If you look at where `runtime.fcntl` is used, the code looks like this:

```go
func Fcntl(fd int, cmd int, arg int) (int, error) {
        val, errno := fcntl(int32(fd), int32(cmd), int32(arg))
        if val == -1 {
                return int(val), syscall.Errno(errno)
        }
        return int(val), nil
}
```

So `val` needs to be -1 when there's an error.
This probably means you can simplify the code like this:

```suggestion
	ret = int32(libc_fcntl(int(fd), int(cmd), int(arg)))
	errno = *libc_errno_location()
	return
```

**Comment by leongross** [comment]
Thanks a lot for reviewing and merging @aykevl

**Pull request merged**